### PR TITLE
Ensure that releasing listener is called

### PR DESCRIPTION
**Pull request by jasontedor**
When sending a response to a client, we attach a releasing listener to the channel promise. If the client disappears before the response is sent, the releasing listener was never notified. The reason the listeners were never notified was due to a mistaken invocation of write and flush on the channel which has two overrides: one that takes an existing promise, and one that does not and instead creates a new promise. When the client disappears, it is this latter promise that is notified, which does not contain the releasing listener. This commit addreses this issue by invoking the override that passes our channel promise through.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpChannel.java [hunk 1]
```diff
@@ -134,11 +134,13 @@ public void sendResponse(RestResponse response) {
...
 
+            final Object msg;
             if (pipelinedRequest != null) {
-                channel.writeAndFlush(pipelinedRequest.createHttpResponse(resp, promise));
+                msg = pipelinedRequest.createHttpResponse(resp);
             } else {
-                channel.writeAndFlush(resp, promise);
+                msg = resp;
             }
+            channel.writeAndFlush(msg, promise);
```
Comment (bleskes) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpChannel.java]: I think the release flag should be move after this? if we don't use the promise, it's useless no?

File: modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpChannelTests.java [hunk 2]
```diff
@@ -217,6 +220,24 @@ public void testHeadersSet() {
...
 
+    public void testReleaseOnSendToClosedChannel() {
+        final Settings settings = Settings.builder().build();
+        final NamedXContentRegistry registry = xContentRegistry();
+        try (Netty4HttpServerTransport httpServerTransport =
+                     new Netty4HttpServerTransport(settings, networkService, bigArrays, threadPool, registry, new NullDispatcher())) {
+            final FullHttpRequest httpRequest = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
+            final EmbeddedChannel embeddedChannel = new EmbeddedChannel();
+            final Netty4HttpRequest request = new Netty4HttpRequest(registry, httpRequest, embeddedChannel);
+            final HttpPipelinedRequest pipelinedRequest = new HttpPipelinedRequest(request.request(), 1);
```
Comment (bleskes) [inline modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpChannelTests.java]: should we randomize if we use pipelining or not?

**Review by bleskes — CHANGES_REQUESTED** [review]
Great catch. LGTM, though I left one important question about the release logic.

File: modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpChannelTests.java (hunk 2 above)
Comment (jasontedor) [inline modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpChannelTests.java]: I pushed 56a9f5e0b79080afd7d29173b44960a01216bef1 for this already. 😉

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpChannel.java (hunk 1 above)
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpChannel.java]: I pushed d2a5ab28f716a8daa21a61decf9fa78c12b0cf07.

**Review by bleskes — APPROVED** [review]
LGTM

**Review by Tim-Brooks — COMMENTED** [review]
Looks like this was just merged, but I did just spend sometime looking at this and everything looked good to me.

**Comment by jasontedor** [comment]
Thanks @bleskes and @tbrooks8.

**Comment by asafb-wix** [comment]
Which versions are affected by this? we are seeing some issues with 5.2.0 that could be explained by this if this issue was present then?

**Comment by jasontedor** [comment]
> Which versions are affected by this?

All versions prior to version 5.2.2.

**Comment by jchipmunk** [comment]
Can you backport the pull request to Elasticsearch 5.1.x?

**Comment by jasontedor** [comment]
> Can you backport the pull request to Elasticsearch 5.1.x?

Sorry, we will not do that as the 5.1 series will not see any additional releases.

**Comment by ebuildy** [comment]
Can this happen on 2.4 also? What do you recommend to avoid this?

**Comment by jasontedor** [comment]
No, this bug was specific to early 5.x releases.

**Comment by ebuildy** [comment]
thanks you, I will post something in discuss so.

**Comment by ustczz** [comment]
@jasontedor hi,thanks for fix the bug.
the bug aslo affect us , then we upgrate through rolling upgrades from 5.0.0 to 5.6.1 without search guid,but the problem Still exists. is not search guid or other method?
thank you!

**Comment by jasontedor** [comment]
@nitishgoyal13 My wild guess is that has nothing to do with this issue and that that’s the accounting breaker. Check your stats, I suspect you’re running at capacity from a large number of shards.

**Pull request merged**